### PR TITLE
[RISCV] Flatten the ImpliedExts table in RISCVISAInfo.cpp

### PR DESCRIPTION
**Pull request by topperc**
Previously we had an individiaul global array of implied extensions for each extension that needed it. This allowed each array to have a different length. Then we had a sorted table that stored pointers and size for the indivual arrays keyed by the extension name.

This patch changes the sorted table to use multiple rows if multiple extensions are implied. We use equal_range instead of lower_bound to find all the rows that apply to a given extension.

The CombineIntoExts array was also modified to store only the extension name that need to be combined. This extension name is looked up in the implied table to find all the extensions it depends on.

Stacked on #89955 because it was easier to flatten the table in tblgen than to do it by hand.

### CODE REVIEW

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 1]
```diff
@@ -880,17 +886,17 @@ void RISCVISAInfo::updateImplication() {
...
-          continue;
-        if (Exts.count(ImpliedExt))
-          continue;
-        auto Version = findDefaultVersion(ImpliedExt);
-        addExtension(ImpliedExt, Version.value());
-        WorkList.insert(ImpliedExt);
-      }
+    auto Range = std::equal_range(std::begin(ImpliedExts),
+                                  std::end(ImpliedExts), ExtName);
+    for (auto I = Range.first, E = Range.second; I != E; ++I) {
```
Comment (preames) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:891]: I think you could use a foreach style loop here?

**Review by preames — APPROVED** [review]
LGTM

File: llvm/lib/TargetParser/RISCVISAInfo.cpp (hunk 1 above)
Comment (topperc) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:891]: std::for_each?

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 2]
```diff
@@ -880,17 +886,17 @@ void RISCVISAInfo::updateImplication() {
...
-      for (const char *ImpliedExt : I->Exts) {
-        if (WorkList.count(ImpliedExt))
-          continue;
-        if (Exts.count(ImpliedExt))
-          continue;
-        auto Version = findDefaultVersion(ImpliedExt);
-        addExtension(ImpliedExt, Version.value());
-        WorkList.insert(ImpliedExt);
-      }
+    auto Range = std::equal_range(std::begin(ImpliedExts),
```
Comment (wangpc-pp) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:889]: We don't have `llvm::equal_range` wrapper? Maybe we should add it.

Comment (4vtomat) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:889]: What's the benefit of wrappers?

Comment (wangpc-pp) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:889]: Just like other wrappers in STLExtras.h, we can save our lives. :-)

Comment (4vtomat) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:889]: Got it, I'm just curious about the advantages of those llvm wrappers.
I was just thinking whether I should use the wrapper or std every time lol~

**Review by asb — APPROVED** [review]
LGTM, thanks again for your work on moving extension definitions to tblgen.

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Pull request merged**